src/codecs/types.ts

```diff
@@ -0,0 +1,77 @@
+import * as stencila from '@stencila/schema'
+import { VFile } from '../util/vfile'
+
+export interface GlobalEncodeOptions<CodecOptions extends object = {}> {
+  format?: string
+  filePath?: string
+  isStandalone?: boolean
+  isBundle?: boolean
+  theme?: 'eLife' | 'stencila'
+  codecOptions?: CodecOptions
+}
+
+/**
+ * The interface for a codec.
+ *
+ * A codec is simply a module with these constants
+ * and functions (some of which are optional).
+ *
+ * Note that our use of the term "codec", is consistent with our usage elsewhere in Stencila
+ * as something that creates or modifies executable document, and
+ * differs from the usage of [`unified`](https://github.com/unifiedjs/unified#processorcodec).
+ */
+export abstract class Codec<
+  EncodeOptions extends object = {},
+  DecodeOptions extends object = {}
+> {
+  /**
+   * An array of [IANA Media Type](https://www.iana.org/assignments/media-types/media-types.xhtml)
+   * that the codec can decode/encode.
+   */
+  public abstract mediaTypes: string[]
```

Comment:
Can this (and other properties) be marked `readonly`?

Reply:
@beneboy unfortunately the answer is no :/
Discussion tracking the issue here: https://github.com/Microsoft/TypeScript/issues/29634

For now I've manually set the `readonly` prop in the derived classes

Reply:
@beneboy unfortunately the answer is no :/
Discussion tracking the issue here: https://github.com/Microsoft/TypeScript/issues/29634

For now I've manually set the `readonly` prop in the derived classes